lock-api/src/main/java/com/palantir/lock/watch/LockWatchEventCacheImpl.java

```diff
@@ -0,0 +1,273 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.watch;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import org.immutables.value.Value;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Ordering;
+import com.google.common.collect.TreeMultimap;
+import com.google.errorprone.annotations.concurrent.GuardedBy;
+import com.palantir.lock.LockDescriptor;
+import com.palantir.lock.v2.LockToken;
+import com.palantir.logsafe.Preconditions;
+
+public final class LockWatchEventCacheImpl implements LockWatchEventCache {
+    private final ClientLockWatchEventLog eventLog;
+    private final HashMap<Long, MapEntry> timestampMap = new HashMap<>();
+    private final TreeMultimap<IdentifiedVersion, Long> aliveVersions =
+            TreeMultimap.create(IdentifiedVersion.comparator(), Ordering.natural());
+
+    @GuardedBy("this")
+    private boolean failed = false;
+
+
+    private LockWatchEventCacheImpl(ClientLockWatchEventLog eventLog) {
+        this.eventLog = eventLog;
+    }
+
+    public static LockWatchEventCacheImpl create() {
+        return create(ClientLockWatchEventLogImpl.create());
+    }
+
+    @VisibleForTesting
+    static LockWatchEventCacheImpl create(ClientLockWatchEventLog eventLog) {
+        return new LockWatchEventCacheImpl(eventLog);
+    }
+
+    @Override
+    public Optional<IdentifiedVersion> lastKnownVersion() {
+        return eventLog.getLatestKnownVersion();
+    }
+
+    @Override
+    public synchronized void processStartTransactionsUpdate(
+            Set<Long> startTimestamps,
+            LockWatchStateUpdate update) {
+        ensureNotFailed(() -> {
+            Optional<IdentifiedVersion> latestVersion = processEventLogUpdate(update);
+
+            latestVersion.ifPresent(
+                    version -> startTimestamps.forEach(timestamp -> {
+                        timestampMap.put(timestamp, MapEntry.of(version));
+                        aliveVersions.put(version, timestamp);
+                    }));
+        });
+    }
+
+    @Override
+    public synchronized void processGetCommitTimestampsUpdate(
+            Collection<TransactionUpdate> transactionUpdates,
+            LockWatchStateUpdate update) {
+        ensureNotFailed(() -> {
+            Optional<IdentifiedVersion> latestVersion = processEventLogUpdate(update);
+
+            latestVersion.ifPresent(version ->
+                    transactionUpdates.forEach(transactionUpdate -> {
+                        MapEntry previousEntry = timestampMap.get(transactionUpdate.startTs());
+                        Preconditions.checkNotNull(previousEntry, "Start timestamp missing from cache");
```

Comment:
This, `getCommitUpdate`, and `getEventsForTransactions` can all fail (and indeed, throw) if a leader election occurs between the start and end of the transaction. Presumably, this is the correct behaviour, but that does require the callers to catch these cases.

Open to ideas for alternatives. `getCommitUpdate` should probably invalidateAll, this method should maybe not fail (it's only used for `getCommitUpdate`), but not sure about what we should do for `getEventsForTransactions`, given that it is intended to be called outside Atlas.

Reply:
As per my other comment, let's review this carefully because a lot of cases you need to recover from by retrying the transactions, so we can't throw Illegal(State|Argument) exceptions.

Reply:
Yeah, I agree.

Reply:
ARGH

cyclic dependency if I want to use `TransactionFailedRetriableException` as `atlasdb-api` depends on `lock-api`, and we would need to depend on `atlasdb-api` to get access to `TransactionFailedRetriableException`...

Reply:
I have a work around but it's not super nice.